glib/src/object.rs

```diff
@@ -3267,6 +3281,75 @@ wrapper! {
     }
 }
 
+type ManuallyDropPinBox<T> = ManuallyDrop<Pin<Box<T>>>;
+
+// rustdoc-stripper-ignore-next
+/// A handle to disconnect a weak ref notify closure.
+pub struct WeakRefNotify<T: ObjectType, F: Fn() + 'static>(
+    ManuallyDropPinBox<(*mut GObject, crate::ThreadGuard<F>, Cell<bool>)>,
+    gobject_ffi::GWeakNotify,
+    PhantomData<*mut T>,
+);
+
+impl<T: ObjectType, F: Fn() + 'static> WeakRefNotify<T, F> {
+    pub(crate) fn new(obj: &T, f: F) -> WeakRefNotify<T, F> {
+        unsafe extern "C" fn notify_func<F: Fn()>(
+            data: ffi::gpointer,
+            _obj: *mut gobject_ffi::GObject,
+        ) {
+            let data = &mut *(data as *mut (*mut GObject, crate::ThreadGuard<F>, Cell<bool>));
+
+            (data.1.get_ref())();
+
+            // Drop the data when the `WeakRefNotify` handle was already dropped
+            if data.2.get() {
+                std::ptr::drop_in_place(data);
+            } else {
+                data.2.set(true);
+            }
+        }
+
+        unsafe {
+            let obj_ptr = obj.as_object_ref().to_glib_none().0;
+            let mut data = ManuallyDrop::new(Box::pin((
+                obj_ptr,
+                crate::ThreadGuard::new(f),
+                Cell::new(false),
+            )));
+            let data_ptr: *mut _ = Pin::as_mut(&mut data).get_unchecked_mut();
+            gobject_ffi::g_object_weak_ref((*data).0, Some(notify_func::<F>), data_ptr as *mut _);
+
+            WeakRefNotify(data, Some(notify_func::<F>), PhantomData)
+        }
+    }
+
+    #[doc(alias = "g_object_weak_unref")]
+    pub fn disconnect(self) {
+        unsafe {
+            let data = &*(self.0);
+            let data_ptr: *const _ = Pin::as_ref(&self.0).get_ref();
+
+            if !data.2.get() {
+                gobject_ffi::g_object_weak_unref(data.0, self.1, data_ptr as *mut _);
+                data.2.set(true);
+            }
+        }
+    }
+}
+impl<T: ObjectType, F: Fn() + 'static> Drop for WeakRefNotify<T, F> {
+    fn drop(&mut self) {
+        let data = &*(self.0);
+        // Drop the data only when the callback was disconnected or when it was already called
+        if data.2.get() {
+            unsafe {
+                ManuallyDrop::drop(&mut self.0);
```

Comment:
This is unsound. If the `WeakRefNotify` is dropped from one thread and ends up in this line, and the original thread ends up in the `notify` function and drops it there another time, you have a double-free.

Reply:
ah true. But `WeakRefNotify` can't be moved to other threads.

Reply:
It should be possible to move it to other threads though :) Nothing wrong with that other than your implementation right now.